rainier-core/src/main/scala/com/stripe/rainier/sampler/Sampler.scala

```diff
@@ -1,13 +1,85 @@
 package com.stripe.rainier.sampler
 
 import com.stripe.rainier.compute._
+import scala.annotation.tailrec
 
 trait Sampler {
   def sample(density: Real, warmupIterations: Int)(
       implicit rng: RNG): Stream[Sample]
 }
 
 final case class Sample(accepted: Boolean, parameters: Array[Double])
+final case class Diagnostics(rHat: Double, effectiveSampleSize: Double)
+
+object Diagnostics {
+  def apply(traces: Seq[Array[Double]]): Diagnostics = {
+    val m = traces.size.toDouble
+    val n = traces.head.size.toDouble
+    val (rHat, v) = rHatAndV(traces, n, m)
+    val ac = autocorrelation(traces, n, m, v, 1, 0.0)
+    val ess = n * m / (1 + (2 * ac))
+    Diagnostics(rHat, ess)
+  }
+
+  def rHatAndV(traces: Seq[Array[Double]],
+               n: Double,
+               m: Double): (Double, Double) = {
+
+    val means = traces.map { t =>
+      t.sum / n
+    }
+
+    val meanMean = means.sum / m
+
+    val b = (n / (m - 1)) * means.map { m =>
+      Math.pow(m - meanMean, 2)
+    }.sum
+
+    val variances =
+      traces.zip(means).map {
+        case (t, m) =>
+          t.map { a =>
+            Math.pow(a - m, 2)
+          }.sum / (n - 1)
+      }
+
+    val w = variances.sum / m
+
+    val v =
+      (n - 1).toDouble / n * w +
+        b / n
+
+    val rHat = Math.sqrt(v / w)
+    (rHat, v)
+  }
+
+  @tailrec
+  def autocorrelation(traces: Seq[Array[Double]],
+                      n: Double,
+                      m: Double,
+                      v: Double,
+                      lag: Int,
+                      acc: Double): Double = {
+    val vt = traces.map { trace =>
+      variogram(trace, lag)
+    }.sum / m
+    val pt = 1.0 - (vt / (2.0 * v))
+    if (pt > 0.0 && lag < 100)
```

Comment:
This probably isn't the right termination criteria. The Stan manual mentions that NUTS never produces negative autocorrelation and so they use it as an indication that noise > signal, but since we're using HMC not NUTS this likely doesn't apply. Not sure what I should use, though.